Provide unit testing devDependencies from wp-scripts

## Problem

`wp-scripts` provides some of the framework necessary to unit-test plugins, but not all. One specific piece that's missing is many of the dev-dependencies still need to be added to the plugin's `package.json`. For example, https://github.com/WordPress/wordcamp.org/pull/211/files.

## Proposed Solution

Those dependencies (and any others) should be added to `wp-scripts`'s `package.json`, so that plugins inherit them automatically.

cc @gziolo , who I think is planning to work on some related issues.

Er, `api-fetch` was a bad example, since that'd probably be better suited for an e2e test, but the same dependency hell comes up when trying to use `i18n` or anything else. I have to:

1. `npm run test source/blocks/schedule/`
1. scroll back to see the `Cannot find module '@wordpress/i18n' ...` error
1. add `@wordpress/i18n` to my `devDependencies`
1. `npm i @wordpress/i18n && npm run test source/blocks/schedule/`
1. repeat those steps for every package, even stuff like `react-autosize-textarea` that I'm not using directly in my component.

--- a/.eslintrc.js
+++ b/.eslintrc.js
@@ -30,6 +30,8 @@ module.exports = {
 		wp: 'off',
 	},
 	rules: {
+		'@wordpress/dependency-group': 'error',
+		'@wordpress/gutenberg-phase': 'error',
 		'@wordpress/react-no-unsafe-timeout': 'error',
 		'no-restricted-syntax': [
 			'error',

--- a/.eslintrc.js
+++ b/.eslintrc.js
@@ -126,7 +126,6 @@ module.exports = {
 		},
 		{
 			files: [
-				'packages/block-serialization-spec-parser/shared-tests.js',
 				'packages/jest*/**/*.js',
 			],
 			extends: [

--- a/.eslintrc.js
+++ b/.eslintrc.js
@@ -20,7 +20,6 @@ module.exports = {
 	root: true,
 	extends: [
 		'plugin:@wordpress/eslint-plugin/recommended',
-		'plugin:jest/recommended',
 		'plugin:eslint-comments/recommended',
 	],
 	plugins: [
@@ -126,15 +125,19 @@ module.exports = {
 			],
 		},
 		{
+			files: [
+				'packages/block-serialization-spec-parser/shared-tests.js',
+				'packages/jest*/**/*.js',
+			],
+			extends: [
+				'plugin:@wordpress/eslint-plugin/test-unit',
+			],
+		},
+		{
 			files: [ 'packages/e2e-test*/**/*.js' ],
-			env: {
-				browser: true,
-			},
-			globals: {
-				browser: 'readonly',
-				page: 'readonly',
-				wp: 'readonly',
-			},
+			extends: [
+				'plugin:@wordpress/eslint-plugin/test-e2e',
+			],
 		},
 	],
 };

--- a/package.json
+++ b/package.json
@@ -101,7 +101,6 @@
 		"enzyme": "3.9.0",
 		"eslint-plugin-eslint-comments": "3.1.2",
 		"eslint-plugin-import": "2.18.2",
-		"eslint-plugin-jest": "22.14.1",
 		"fast-glob": "2.2.7",
 		"fbjs": "0.8.17",
 		"fs-extra": "8.0.1",

--- a/packages/babel-preset-default/CHANGELOG.md
+++ b/packages/babel-preset-default/CHANGELOG.md
@@ -1,3 +1,9 @@
+## Master
+
+### Bug Fixes
+
+- Added missing `@wordpress/element` dependency which is used internally.
+
 ## 4.4.0 (2019-08-05)
 
 ### Bug Fixes

--- a/packages/babel-preset-default/package.json
+++ b/packages/babel-preset-default/package.json
@@ -36,10 +36,8 @@
 		"@babel/runtime": "^7.4.4",
 		"@wordpress/babel-plugin-import-jsx-pragma": "file:../babel-plugin-import-jsx-pragma",
 		"@wordpress/browserslist-config": "file:../browserslist-config",
+		"@wordpress/element": "file:../element",
 		"core-js": "^3.1.4"
-	},
-	"peerDependencies": {
-		"@babel/core": "^7.0.0"
 	},
 	"publishConfig": {
 		"access": "public"

--- a/packages/block-serialization-spec-parser/.eslintrc.json
+++ b/packages/block-serialization-spec-parser/.eslintrc.json
@@ -1,0 +1,10 @@
+{
+	"overrides": [
+		{
+			"files": [ "shared-tests.js" ],
+			"extends": [
+				"plugin:@wordpress/eslint-plugin/test-unit"
+			]
+		}
+	]
+}

--- a/packages/eslint-plugin/CHANGELOG.md
+++ b/packages/eslint-plugin/CHANGELOG.md
@@ -10,6 +10,10 @@
 - New Rule: [`@wordpress/no-unguarded-get-range-at`](https://github.com/WordPress/gutenberg/blob/master/packages/eslint-plugin/docs/rules/no-unguarded-get-range-at.md)
 - Enable `wp` global by default in the `recommended` config.
 
+### Enhancements
+
+- Remove `@wordpress/dependency-group` and `@wordpress/gutenberg-phase` rules from the `custom` and `recommended` configs and leave them as opt-in features.
+
 ## 2.4.0 (2019-08-05)
 
 ### New Features

--- a/packages/eslint-plugin/CHANGELOG.md
+++ b/packages/eslint-plugin/CHANGELOG.md
@@ -4,11 +4,14 @@
 
 - The [`@wordpress/no-unused-vars-before-return` rule](https://github.com/WordPress/gutenberg/blob/master/packages/eslint-plugin/docs/rules/no-unused-vars-before-return.md) has been improved to exempt object destructuring only if destructuring to more than one property.
 - Stricter JSDoc linting using [`eslint-plugin-jsdoc`](https://github.com/gajus/eslint-plugin-jsdoc).
+- Stricter validation enabled for test files only using new `test-e2e` and `test-unit` rulesets.
 
 ### New Features
 
 - New Rule: [`@wordpress/no-unguarded-get-range-at`](https://github.com/WordPress/gutenberg/blob/master/packages/eslint-plugin/docs/rules/no-unguarded-get-range-at.md)
 - Enable `wp` global by default in the `recommended` config.
+- New ruleset `test-e2e` added for end-to-end tests validation.
+- New ruleset `test-unit` added for unit tests validation.
 
 ### Enhancements
 

--- a/packages/eslint-plugin/README.md
+++ b/packages/eslint-plugin/README.md
@@ -28,10 +28,15 @@ The `recommended` preset will include rules governing an ES2015+ environment, an
 
 Alternatively, you can opt-in to only the more granular rulesets offered by the plugin. These include:
 
+- `custom`
 - `es5`
 - `esnext`
+- `jsdoc`
+- `jshint` (legacy)
 - `jsx-a11y`
 - `react`
+- `test-e2e`
+- `test-unit`
 
 For example, if your project does not use React, you could consider extending including only the ESNext rules in your project using the following `extends` definition:
 

--- a/packages/eslint-plugin/configs/custom.js
+++ b/packages/eslint-plugin/configs/custom.js
@@ -3,8 +3,6 @@ module.exports = {
 		'@wordpress',
 	],
 	rules: {
-		'@wordpress/dependency-group': 'error',
-		'@wordpress/gutenberg-phase': 'error',
 		'@wordpress/no-unused-vars-before-return': 'error',
 		'@wordpress/valid-sprintf': 'error',
 		'@wordpress/no-base-control-with-label-without-id': 'error',

--- a/packages/eslint-plugin/configs/recommended.js
+++ b/packages/eslint-plugin/configs/recommended.js
@@ -10,8 +10,28 @@ module.exports = {
 		node: true,
 	},
 	globals: {
-		window: true,
-		document: true,
+		window: 'readonly',
+		document: 'readonly',
 		wp: 'readonly',
 	},
+	overrides: [
+		{
+			files: [
+				'**/@(test|__tests__)/**/*.js',
+				'**/?(*.)test.js',
+			],
+			extends: [
+				require.resolve( './test-unit.js' ),
+			],
+		},
+		{
+			files: [
+				'**/specs/**/*.js',
+				'**/?(*.)spec.js',
+			],
+			extends: [
+				require.resolve( './test-e2e.js' ),
+			],
+		},
+	],
 };

--- a/packages/eslint-plugin/configs/recommended.js
+++ b/packages/eslint-plugin/configs/recommended.js
@@ -10,12 +10,13 @@ module.exports = {
 		node: true,
 	},
 	globals: {
-		window: 'readonly',
-		document: 'readonly',
+		window: true,
+		document: true,
 		wp: 'readonly',
 	},
 	overrides: [
 		{
+			// Unit test files and their helpers only.
 			files: [
 				'**/@(test|__tests__)/**/*.js',
 				'**/?(*.)test.js',
@@ -25,6 +26,7 @@ module.exports = {
 			],
 		},
 		{
+			// End-to-end test files and their helpers only.
 			files: [
 				'**/specs/**/*.js',
 				'**/?(*.)spec.js',

--- a/packages/eslint-plugin/configs/test-e2e.js
+++ b/packages/eslint-plugin/configs/test-e2e.js
@@ -1,0 +1,13 @@
+module.exports = {
+	extends: [
+		'plugin:jest/recommended',
+	],
+	env: {
+		browser: true,
+	},
+	globals: {
+		browser: 'readonly',
+		page: 'readonly',
+		wp: 'readonly',
+	},
+};

--- a/packages/eslint-plugin/configs/test-unit.js
+++ b/packages/eslint-plugin/configs/test-unit.js
@@ -1,0 +1,5 @@
+module.exports = {
+	extends: [
+		'plugin:jest/recommended',
+	],
+};

--- a/packages/eslint-plugin/package.json
+++ b/packages/eslint-plugin/package.json
@@ -25,6 +25,7 @@
 	"main": "index.js",
 	"dependencies": {
 		"babel-eslint": "^10.0.2",
+		"eslint-plugin-jest": "^22.15.1",
 		"eslint-plugin-jsdoc": "^15.8.0",
 		"eslint-plugin-jsx-a11y": "^6.2.3",
 		"eslint-plugin-react": "^7.14.3",

--- a/packages/is-shallow-equal/.eslintrc.json
+++ b/packages/is-shallow-equal/.eslintrc.json
@@ -1,7 +1,17 @@
 {
 	"root": true,
-	"extends": "plugin:@wordpress/eslint-plugin/es5",
+	"extends": [
+		"plugin:@wordpress/eslint-plugin/es5"
+	],
 	"env": {
 		"node": true
-	}
+	},
+	"overrides": [
+		{
+			"files": [ "@(test|benchmark)/*.js" ],
+			"extends": [
+				"plugin:@wordpress/eslint-plugin/recommended"
+			]
+		}
+	]
 }

--- a/packages/is-shallow-equal/benchmark/.eslintrc.json
+++ b/packages/is-shallow-equal/benchmark/.eslintrc.json
@@ -1,4 +1,0 @@
-{
-	"root": true,
-	"extends": "../../../.eslintrc.js"
-}

--- a/packages/is-shallow-equal/test/.eslintrc.json
+++ b/packages/is-shallow-equal/test/.eslintrc.json
@@ -1,4 +1,0 @@
-{
-	"root": true,
-	"extends": "../../../.eslintrc.js"
-}

--- a/packages/scripts/CHANGELOG.md
+++ b/packages/scripts/CHANGELOG.md
@@ -5,6 +5,7 @@
 - The bundled `puppeteer` dependency has been updated from requiring `1.6.1` to requiring `^1.19.0` ([#16875](https://github.com/WordPress/gutenberg/pull/16875)). It uses Chromium v77 instead of Chromium v69.
 - The bundled `jest-puppeteer` dependency has been updated from requiring `^4.0.0` to requiring `^4.3.0` ([#16875](https://github.com/WordPress/gutenberg/pull/16875)).
 - The bundled `eslint` dependency has been updated from requiring `^5.16.0` to requiring `^6.1.0`.
+- The bundled `@wordpress/eslint-plugin` dependency has been updated to the next major version `^3.0.0` due to new ESLint rules enabled for all test files.
 
 ## 3.4.0 (2019-08-05)
 

